src/url-params.ts

```diff
@@ -0,0 +1,111 @@
+import * as _ from 'lodash';
+import {
+	getCookie,
+	setCookie,
+	hasCookieSupport,
+} from '@analytics/cookie-utils';
+import { Mixpanel } from 'mixpanel-browser';
+
+const URL_PARAM_DEVICE_ID = 'd_id';
+const COOKIES_DEVICE_IDS = '__analytics_dids';
+
+const deviceIdSeparator = /\s*,\s*/;
+
+const cookiesSupported = hasCookieSupport();
+
+/**
+ * AnalyticsUrlParams helps with handling analytics-related URL parameters
+ * like anonymous device identifier.
+ */
+export class AnalyticsUrlParams {
+	private deviceIds: string | null;
+
+	constructor(private mixpanel?: Mixpanel) {}
+
+	private storeInboundDeviceIds(ids: string, currentDeviceId: string | null) {
+		const list = ids ? ids.split(deviceIdSeparator) : [];
+		if (currentDeviceId) {
+			list.push(currentDeviceId);
+		}
+		const res = this.retrieveAndMerge(list);
+		this.deviceIds = res;
+		if (cookiesSupported) {
+			setCookie(COOKIES_DEVICE_IDS, res);
+		}
+		return res;
+	}
+
+	private retrieveAndMerge(input: string[]) {
+		const storedValue = cookiesSupported ? getCookie(COOKIES_DEVICE_IDS) : null;
+		const storedList = storedValue ? storedValue.split(deviceIdSeparator) : [];
+		return _.union(storedList, input).join(',');
+	}
+
+	private takeFirstId(passedId: string) {
+		return passedId.split(deviceIdSeparator, 1)[0];
+	}
+
+	static clearCookies() {
+		if (cookiesSupported) {
+			setCookie(COOKIES_DEVICE_IDS, '');
+		}
+	}
+
+	/**
+	 * Analyzes the query string and stores the anonymous device ID if it's there.
+	 * Returns null if the current URL query string should not be modified after consuming the data.
+	 *
+	 * @param queryString URL query string to process
+	 * @return a modified query string that hides processed/consumed parameters if the query string should be modified
+	 */
+	consumeUrlParameters(queryString: string): string | null {
+		const params = new URLSearchParams(queryString);
+
+		const passedDeviceId = params.get(URL_PARAM_DEVICE_ID);
+		if (passedDeviceId) {
+			let originalMpId: string | null = null;
+
+			if (this.mixpanel != null) {
+				originalMpId = this.mixpanel.get_distinct_id();
+
+				// Switch mixpanel ID to using the passed device ID, so we can track events in one timeline branch.
+				// Previous user activity recorded with another device ID will be merged upon signup,
+				// since we store the previous ID to pass it to other sites.
+				const mpId = this.takeFirstId(passedDeviceId);
+				this.mixpanel.register({
+					distinct_id: mpId,
+					$device_id: mpId,
+				});
+			}
+
+			this.storeInboundDeviceIds(passedDeviceId, originalMpId);
+
+			params.delete(URL_PARAM_DEVICE_ID);
+			return params.toString();
+		} else {
+			return null;
+		}
+	}
+
+	/**
+	 * @return all anonymous device IDs that can be passed to other sites, separated with comma
+	 */
+	allDeviceIds(): string {
```

Comment:
I think it is better to return a `string[]` instead of the already concatenated string. You can have a separate method that will do the concatenation, otherwise the name is confusing (judging by the name, I would expect an array).

Reply:
Yep, I agree.
this method now returns an array